pkg/dashboard/manager.go

```diff
@@ -0,0 +1,215 @@
+	members  []*pdpb.Member
+}
+
+// NewManager creates a new Manager.
+func NewManager(srv *server.Server, s *apiserver.Service, redirector *Redirector) *Manager {
+	ctx, cancel := context.WithCancel(srv.Context())
+	return &Manager{
+		ctx:           ctx,
+		cancel:        cancel,
+		srv:           srv,
+		service:       s,
+		redirector:    redirector,
+		enableDynamic: srv.GetConfig().EnableDynamicConfig,
+	}
+}
+
+func (m *Manager) start() {
+	m.wg.Add(1)
+	go m.serviceLoop()
+}
+
+func (m *Manager) stop() {
+	m.cancel()
+	m.wg.Wait()
+	log.Info("exit dashboard loop")
+}
+
+func (m *Manager) serviceLoop() {
+	defer logutil.LogPanic()
+	defer m.wg.Done()
+
+	ticker := time.NewTicker(checkInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-m.ctx.Done():
+			m.stopService()
+			return
+		case <-ticker.C:
+			m.updateInfo()
+			m.checkAddress()
+		}
+	}
+}
+
+// updateInfo updates information from the server.
+func (m *Manager) updateInfo() {
+	if !m.srv.GetMember().IsLeader() {
+		m.isLeader = false
+		m.rc = nil
+		m.members = nil
+		return
+	}
+
+	m.isLeader = true
+	m.rc = m.srv.GetRaftCluster()
+	if m.rc == nil || !m.rc.IsRunning() {
+		m.members = nil
+		return
+	}
+
+	var err error
+	if m.members, err = cluster.GetMembers(m.srv.GetClient()); err != nil {
+		log.Error("failed to get members")
+	}
+}
+
+// checkDashboardAddress checks if the dashboard service needs to change due to dashboard address is changed.
+func (m *Manager) checkAddress() {
+	dashboardAddress := m.srv.GetScheduleOption().GetDashboardAddress()
+	switch dashboardAddress {
+	case "auto":
+		if m.isLeader && len(m.members) > 0 {
+			m.setNewAddress()
+		}
+		return
+	case "none":
+		m.redirector.SetAddress("")
+		m.stopService()
+		return
+	default:
+		if _, err := url.Parse(dashboardAddress); err != nil {
+			log.Error("illegal dashboard address", zap.String("address", dashboardAddress))
+			return
+		}
+
+		if m.isLeader && m.needResetAddress(dashboardAddress) {
+			m.setNewAddress()
+			return
+		}
+	}
+
+	m.redirector.SetAddress(dashboardAddress)
+
+	clientUrls := m.srv.GetMemberInfo().GetClientUrls()
+	if len(clientUrls) > 0 && clientUrls[0] == dashboardAddress {
+		m.startService()
+	} else {
+		m.stopService()
+	}
+}
+
+func (m *Manager) needResetAddress(addr string) bool {
+	if len(m.members) == 0 {
+		return false
+	}
+
+	for _, member := range m.members {
+		if member.GetClientUrls()[0] == addr {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (m *Manager) setNewAddress() {
+	// get new dashboard address
+	members := m.members
+	var addr string
+	switch len(members) {
+	case 1:
+		addr = members[0].GetClientUrls()[0]
+	default:
+		addr = members[0].GetClientUrls()[0]
+		leaderID := m.srv.GetMemberInfo().MemberId
+		sort.Slice(members, func(i, j int) bool { return members[i].GetMemberId() < members[j].GetMemberId() })
+		for _, member := range members {
+			if member.MemberId != leaderID {
+				addr = member.GetClientUrls()[0]
```

Comment:
what if the join a new member or leader changed?

Reply:
Nothing happens unless the member that is the dashboardAddress is deleted.